### PR TITLE
feat: add core device moninj proxy

### PR DESCRIPTION
**Pull request by stevefan1999**
# ARTIQ Pull Request

## Description of Changes

This adds a core device proxy controller which is split into two parts, for data synchronization that listens on 2383. This PR assumes the controller can reach core device, i.e. within the same network. 

This has the benefits of not only consolidating all core device communication into one which can save bandwidths, but it also cater to some restricted network conditions such as ARTIQ dashboard trying to communicate with core device in different LAN/separated network. It is possible that master/dashboard/core device all ran in different networks.

### Related Issue

This PR is part of resolving #1435

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [ ] Close/update issues.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [ ] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/coredevice/comm_moninj.py [hunk 1]
```diff
@@ -21,6 +17,16 @@ class TTLOverride(Enum):
...
+class ICommMonInj:
+    def monitor_probe(self, enable: int, channel: int, probe: int): raise NotImplementedError
+
+    def monitor_injection(self, enable: int, channel: int, overrd: int): raise NotImplementedError
+
+    def inject(self, channel: int, override: int, value: int): raise NotImplementedError
+
+    def get_injection_status(self, channel: int, override: int): raise NotImplementedError
+
+
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_moninj.py:29]: See comments on previous PR, you most likely don't need to change this file at all and again, type annotations if you want them should be in a separate PR, applied to a whole module at a time, and checked for correctness.

**Comment by sbourdeauducq** [comment]
Controllers are normally called ``aqctl_*``
The documentation needs an update (port list, release notes, and maybe other places).
``artiq_ddb_template`` should generate an entry for the proxy since it's wanted in most cases.

**Comment by stevefan1999** [comment]
@sbourdeauducq any problem regarding it together with #1753?

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    level=logging.DEBUG,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[
+        logging.StreamHandler()
+    ]
+)
+logger = logging.getLogger(__name__)
+
+
+class Connectable(ABC):
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:27]: Nothing else is using ABC in ARTIQ and iirc it's not very useful - remove.

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    def __init__(self):
+        self.connected_event = asyncio.Event()
+
+    @abstractmethod
+    async def connect(self, addr, port): pass
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:32]: Usual coding pattern in ARTIQ is ``raise NotImplementedError`` in this case.

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+class HasSupervisor(ABC):
+    @property
+    @abstractmethod
+    def supervisor(self): pass
+
+
+class HasCoreAddr(ABC):
+    @property
+    @abstractmethod
+    def core_addr(self): pass
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:57]: Are those little classes really useful?

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    async def connect(self):
+        self.activation_event.set()
+        await self.master.connect(self.master_addr, self.master_notify_port)
+
+        # while self.activation_event.is_set() and not self.master.connected:
+        #     try:
+        #         await self.master.connect(self.master_addr, self.master_notify_port)
+        #     except:
+        #         logging.error("master still unreachable, retrying...")
+        #         await asyncio.sleep(10)
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:225]: remove

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+class MonInjSupervisor:
+    def __init__(self, master, notify_port):
+        self.notify = Notifier(dict({
+            "monitor": dict(),
+            "injection_status": dict(),
+            "connected": {"coredev": False, "master": False}
+        }))
+        self.master_addr = master
+        self.master_notify_port = notify_port
+        self.core_, self.master_ = MonInjCoredev(self), MonInjMaster(self)
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:205]: Why the underscore?

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+from artiq import __version__ as artiq_version
+from artiq.coredevice.comm_moninj import CommMonInj
+
+logging.basicConfig(
+    level=logging.DEBUG,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[
+        logging.StreamHandler()
+    ]
+)
```
Comment (sbourdeauducq) [inline artiq/frontend/aqctl_proxy_moninj.py:23]: Why?
See other code.

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 4 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py:57]: It is useful that the event handlers can access the properties of "child class". My intention is to make it look like a Rust trait (also to emulate composition over inheritance), but it seems like there is no way to do that in Python elegantly.

Comment (sbourdeauducq) [reply artiq/frontend/aqctl_proxy_moninj.py:57]: Please just keep it simple - this is supposed to be just a little program, no need to invent new Python programming techniques here.

File: artiq/coredevice/comm_moninj.py [hunk 8]
```diff
@@ -32,7 +32,12 @@ def __init__(self, monitor_cb, injection_status_cb, disconnect_cb=None):
         self._receive_task = None
 
     async def connect(self, host, port=1383):
+        if self._writer:
+            await self.close()
+
         self._reader, self._writer = await asyncio.open_connection(host, port)
+        set_keepalive(self._writer.transport._sock, 1, 1, 3)
```
Comment (sbourdeauducq) [inline artiq/coredevice/comm_moninj.py:39]: hack?

Comment (stevefan1999) [reply artiq/coredevice/comm_moninj.py:39]: although it works, yes this is a hack, since _sock is a private field of BaseTransport, which you shouldn't access outside of the class.

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 4 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: Still, this cleanly separate the callback mixin class while still having access to parent class. I'd suggest keeping it.

File: artiq/frontend/artiq_ddb_template.py [hunk 9]
```diff
@@ -38,6 +38,16 @@ def process_header(output, description):
...
                 "command": "aqctl_corelog -p {{port}} --bind {{bind}} " + core_addr
             }},
+            "moninj": {{
+                "type": "controller",
+                "host": "{moninj_host}",
+                "port": "@rpc_port",
+                "master_addr": "{master_addr}",
+                "rpc_port": "{rpc_port}",
+                "pubsub_port": "{pubsub_port}",
+                "command": "aqctl_proxy_moninj --master-addr {{master_addr}} --bind {{bind}} "
```
Comment (occheung) [inline artiq/frontend/artiq_ddb_template.py:48]:
```
$ aqctl_proxy_moninj
aqctl_proxy_moninj: command not found
```
Need tweak in `setup.py` to make it callable?

File: doc/manual/default_network_ports.rst [hunk 10]
```diff
@@ -12,6 +12,10 @@ Default network ports
 +---------------------------------+--------------+
 | Core device (mon/inj)           | 1383         |
 +---------------------------------+--------------+
+| Coredev mon/inj proxy (pubusb)  | 2383         |
```
Comment (occheung) [inline doc/manual/default_network_ports.rst:15]: You meant `pubsub` right?

File: artiq/frontend/artiq_ddb_template.py (hunk 9 above)
Comment (stevefan1999) [reply artiq/frontend/artiq_ddb_template.py:48]: Thanks, I wasn't aware of that

File: artiq/coredevice/coredevice_generic.schema.json [hunk 11]
```diff
@@ -98,7 +120,7 @@
             }
         }
     },
-    "required": ["target", "variant", "hw_rev", "base", "peripherals"],
+    "required": ["target", "variant", "hw_rev", "base", "peripherals", "moninj"],
```
Comment (occheung) [inline artiq/coredevice/coredevice_generic.schema.json:123]: Is there a need for `moninj` entry to be required? Like does it make sense to use localhost for default address and the default ports stated in the manual/`aqctl_proxy_moninj`?

So far it will break all the existing json configurations, though adding an `moninj` entry should be doable.

Comment (stevefan1999) [reply artiq/coredevice/coredevice_generic.schema.json:123]: either we set moninj entries with default values, or the dashboard won't work anymore. iirc it is decided to use moninj proxy for all dashboard interaction so this is pretty much necessary right now

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+    async def on_connected(self):
+        if self.connected:
+            return
+
+        await super().on_connected()
+        logger.info("connected to master")
+        self._supervisor.notify["connected"]["master"] = True
+
+        # if the core device address is not the same of what we memoized, notify the change
+        if hash(self._core_addr_cache) ^ hash(self.core_addr):
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:141]: Is there a particular need to use `hash()` here?
As far as I can tell, it only compares the core address.

Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py:141]: you can't do xor with strings, so i used hash instead

Comment (occheung) [reply artiq/frontend/aqctl_proxy_moninj.py:141]: Just do string comparison like `!=`?

Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: oh yes, i made it overcomplicated

**Comment by stevefan1999** [comment]
the proxy is now in-place usable -- theoretically there is no code changes needed in dashboard unlike what I previous did in #1753, but the communicator needs to point from core device to the proxy. As such, #1753 is now focused on restructuring, with some quality of life changes to cater the upcoming Urukul moninj and future extension by putting back some code into a separate widget file with individual rather than garbling the responsibilities all in one common moninj.py file (with the Urukul patch coming in, that file will contain over 1000 lines -- not very good from a maintenance standpoint).

Of course, currently there is one problem where there are duplicated code in the proxy and the CommMonInj (like the packet format/serialization), so I would like to put them in a common place later on.

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+class MonInjProxy(AsyncioServer):
+    def __init__(self, master, notify_port):
+        super().__init__()
+        self.master_addr = master
+        self.master_notify_port = notify_port
+        self.core, self.master = MonInjCoredev(self), MonInjMaster(self)
+        self.active = False
+        self.mon_fields = MonitoredField()
+        self.clients = []
+        self.endian = ""
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:178]: AFAICT (barely) you just funnel the endianness from comm in `on_connected`.
Not sure if we can just directly access it through `self.core.comm.endian` in `_handle_connection_cr` instead. Though this looks a bit messy as well.

File: doc/manual/getting_started_mgmt.rst [hunk 14]
```diff
@@ -46,6 +46,30 @@ The dashboard should display the list of experiments from the repository folder
 
 .. note:: Multiple clients may be connected at the same time, possibly on different machines, and will be synchronized. See the ``-s`` option of ``artiq_dashboard`` and the ``--bind`` option of ``artiq_master`` to use the network. Both IPv4 and IPv6 are supported.
 
+(Optional) Using a monitor/injection proxy for core device
+----------------------------------------------------------
+
+If you do not direct access to the subnet of control plane, e.g. not having symmetrical access to core device
```
Comment (occheung) [inline doc/manual/getting_started_mgmt.rst:52]: You meant "do not **have** direct access"?

**Review by occheung — COMMENTED** [review]
Probably need some `flake8` after all changes are made.
Will test soon.

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 13 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: done

File: doc/manual/getting_started_mgmt.rst (hunk 14 above)
Comment (stevefan1999) [reply doc/manual/getting_started_mgmt.rst:52]: oops, it's a typo, thanks

File: doc/manual/getting_started_mgmt.rst [hunk 15]
```diff
@@ -46,6 +46,30 @@ The dashboard should display the list of experiments from the repository folder
 
 .. note:: Multiple clients may be connected at the same time, possibly on different machines, and will be synchronized. See the ``-s`` option of ``artiq_dashboard`` and the ``--bind`` option of ``artiq_master`` to use the network. Both IPv4 and IPv6 are supported.
 
+(Optional) Using a monitor/injection proxy for core device
```
Comment (occheung) [inline doc/manual/getting_started_mgmt.rst:49]: Seems not optional when using with dashboard.
Just make sure the doc clearly states that you need to start the proxy. Should just be a doc issue if it is intended.
@sbourdeauducq

File: artiq/dashboard/moninj.py [hunk 16]
```diff
@@ -398,7 +398,7 @@ async def core_connector(self):
             new_core_connection = CommMonInj(self.monitor_cb, self.injection_status_cb,
                     self.disconnect_cb)
             try:
-                await new_core_connection.connect(self.core_addr, 1383)
+                await new_core_connection.connect("localhost", 2383)
```
Comment (occheung) [inline artiq/dashboard/moninj.py:401]: I swear this wasn't here this morning...
@stevefan1999 maybe refrain from force pushing for a while, it is easy to miss these stuff.

Comment (stevefan1999) [reply artiq/dashboard/moninj.py]: this is unintended

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+from sipyco import common_args, pyon
+from sipyco.asyncio_tools import atexit_register_coroutine, AsyncioServer
+from sipyco.packed_exceptions import current_exc_packed
+from sipyco.sync_struct import Subscriber
+
+from artiq import __version__ as artiq_version
+from artiq.coredevice.comm_moninj import CommMonInj
+
+logging.basicConfig(level=logging.INFO)
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:19]: This line seems to have fixed the logging level.

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+            writer.close()
+
+
+def get_argparser():
+    parser = argparse.ArgumentParser(
+        description="ARTIQ Core Device Monitor/Injection Proxy")
+    parser.add_argument("--version", action="version",
+                        version="ARTIQ v{}".format(artiq_version),
+                        help="print the ARTIQ version number")
+    common_args.verbosity_args(parser)
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:369]: Verbosity args seems unused anywhere.

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+
+class MonInjCoredev:
+    def __init__(self, proxy):
+        self.proxy = proxy
+        self.comm = None
+        self._connected = False
+
+    async def connect(self, addr, port):
+        try:
+            logging.info(f"trying to connect to coredev at {addr}:{port}")
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:55]: Just use the same `logger`?

File: artiq/frontend/aqctl_proxy_moninj.py [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+        await self.ddb_notify.close()
+        await self.on_disconnected()
+
+    @property
+    def core_addr(self):
+        return self._core_addr
+
+    async def set_core_addr(self, value):
+        if self._core_addr and value and self._core_addr != value:
+            logging.debug(
```
Comment (occheung) [inline artiq/frontend/aqctl_proxy_moninj.py:146]: Same as above.

File: doc/manual/getting_started_mgmt.rst [hunk 21]
```diff
@@ -46,6 +46,30 @@ The dashboard should display the list of experiments from the repository folder
...
+
+    $ aqctl_proxy_moninj --master-addr <master address> --port <proxy port> --bind <bind address>
+
+And add an entry of moninj to device database: ::
+
+    device_db['moninj'] = {
+        "type": "controller",
+        "host": "::1",
+        "port": "2383",
+        "master_addr": "localhost",
```
Comment (occheung) [inline doc/manual/getting_started_mgmt.rst:65]: **_Maybe_** replace `localhost` with `::1`, just to make it less confusing.

**Review by occheung — COMMENTED** [review]
Verbosity parameter seems not working. See other `frontend` tools.

File: doc/manual/getting_started_mgmt.rst (hunk 21 above)
Comment (occheung) [reply doc/manual/getting_started_mgmt.rst:65]: Actually, you mentioned the possibility of connecting other network already, maybe you should instead parametrize the entire `moninj` entry. A default `moninj` entry is already generated from `artiq_ddb_template`.

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 17 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: fixed

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 18 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py:368]: fixed

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 19 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: fixed

File: artiq/frontend/aqctl_proxy_moninj.py (hunk 20 above)
Comment (stevefan1999) [reply artiq/frontend/aqctl_proxy_moninj.py]: fixed

File: doc/manual/getting_started_mgmt.rst (hunk 21 above)
Comment (stevefan1999) [reply doc/manual/getting_started_mgmt.rst:65]: This is needed for existing users who are upgrading from ARTIQ 6 Iirc

File: artiq/dashboard/moninj.py (hunk 16 above)
Comment (occheung) [reply artiq/dashboard/moninj.py]: Ok, so from what I see this line's equivalent is in #1753, i.e. the proxy is not used until #1753. But its title literally "Refactor", which shouldn't justify enabling a feature.

I am raising it here because I am confused why the proxy enabled in a separate PR that intends to refactor only, when it is implemented in this PR.

**Comment by occheung** [comment]
Confusion on enabling (or its lack of) the usage of proxy moninj on dashboard stated above.
May need doc tweaks depending on what we want to achieve from this PR.

**Review by occheung — APPROVED** [review]
LGTM.
